File: reference/gestures/skip.py
import time


class SkipDetector:
    """Presence gate: are the legs still skipping in place?

    Counts a "step" each time either knee rises above baseline by
    `lift_thresh` (scaled by shoulder width) then comes back down. If no step
    happens within `timeout` seconds, `running` goes False.
    """

    def __init__(self, lift_thresh, timeout):
        self.lift_thresh = lift_thresh
        self.timeout = timeout
        self._last_step = 0.0
        self._up = {"l": False, "r": False}
        self.step_count = 0

    def update(self, f, baseline, now=None):
        now = now if now is not None else time.monotonic()
        lift = self.lift_thresh * baseline.shoulder_width
        rearm = 0.5 * lift  # hysteresis before another step counts

        for side, knee_y in (("l", f.knee_l_y), ("r", f.knee_r_y)):
            rise = baseline.knee_y - knee_y  # y grows down -> rise is positive up
            if not self._up[side] and rise > lift:
                self._up[side] = True
                self._last_step = now
                self.step_count += 1
            elif self._up[side] and rise < rearm:
                self._up[side] = False

        return (now - self._last_step) <= self.timeout

    @property
    def running(self):
        return (time.monotonic() - self._last_step) <= self.timeout

File: src/shared/gestures/skip.ts
import type { Baseline } from "../calibrate";
import type { Features } from "../pose";

/**
 * Presence gate: are the legs still skipping in place? Counts a "step" each
 * time either knee rises above baseline by `liftThresh` (scaled by shoulder
 * width) then comes back down. No step within `timeout` seconds -> not running.
 */
export class SkipDetector {
  private lastStep = -1e9;
  private up = { l: false, r: false };
  stepCount = 0;

  constructor(
    private liftThresh: number,
    private timeout: number,
  ) {}

  update(f: Features, baseline: Baseline, now: number): boolean {
    const lift = this.liftThresh * baseline.shoulderWidth;
    const rearm = 0.5 * lift;

    const knees: [keyof typeof this.up, number][] = [
      ["l", f.kneeLY],
      ["r", f.kneeRY],
    ];
    for (const [side, kneeY] of knees) {
      const rise = baseline.kneeY - kneeY; // y grows down -> positive = up
      if (!this.up[side] && rise > lift) {
        this.up[side] = true;
        this.lastStep = now;
        this.stepCount += 1;
      } else if (this.up[side] && rise < rearm) {
        this.up[side] = false;
      }
    }
    return now - this.lastStep <= this.timeout;
  }
}
